trafix-codec/src/encoder/mod.rs

```diff
@@ -0,0 +1,235 @@
+//! Implementation of the Message encoder.
+
+use bytes::{BufMut, Bytes, BytesMut};
+
+use crate::message::{Body, Header, field::Field};
+
+/// Computes the running FIX checksum (tag 10) while encoding.
+#[derive(Default)]
+struct Digest {
+    checksum: u8,
+}
+
+impl Digest {
+    /// Updates the running checksum using the contents of a [`BytesMut`].
+    ///
+    /// This performs modulo-256 addition across all bytes, matching the FIX
+    /// checksum algorithm.
+    pub fn push(&mut self, input: &BytesMut) {
+        for &b in input.as_ref() {
+            self.checksum = self.checksum.wrapping_add(b);
+        }
+    }
+}
+
+/// ASCII SOH delimiter (0x01) used as field terminator in FIX messages.
+const SOH: u8 = b'\x01';
+
+/// Encodes a full FIX message (header + body + trailer) into a final wire-format `Bytes` buffer
+/// during which fields `BodyLength` and `Checksum` are calculated and set.
+pub(crate) fn encode(header: &Header, body: &Body) -> Bytes {
+    // encode optional headers and the body
+    let regular_fields = encode_regular_fields(header, body);
+
+    // encode 8-BeginString, 9-BodyLength fields
+    let message = encode_framing_headers(header, &regular_fields);
+
+    // encode 10-Checksum field (returning frozen message)
+    encode_checksum(message)
+}
+
+/// Encodes all regular fields (`MsgType`, optional header fields, body fields)
+/// starting at tag 35 and ending before tag 10.
+#[must_use]
+fn encode_regular_fields(header: &Header, body: &Body) -> BytesMut {
+    // reserving the capacity, counting that each field has AT LEAST 4 bytes b"X=Y\x01" to
+    // reduce the number of resizings. We can safely assume that the average number of bytes
+    // per field is around 15 bytes as per our measurements.
+    //
+    // +1 represents the MsgType that's outside the fields vec
+    let mut message = BytesMut::with_capacity((header.fields.len() + body.fields.len() + 1) * 15);
+
+    // MsgType with included SOH char
+    message.extend_from_slice(
+        Field::Custom {
+            tag: 35,
+            value: header.msg_type.into(),
+        }
+        .encode()
+        .as_ref(),
+    );
+    message.put_u8(SOH);
+
+    // Optional header fields
+    for field in &header.fields {
+        // field with included SOH char.. x=ab\x01
+        let mut field_soh = field.encode();
+        field_soh.push(SOH);
+
+        // encode the field into the message
+        message.extend_from_slice(field_soh.as_ref());
+    }
+
+    // Body fields
+    for field in &body.fields {
+        // field with included SOH char.. x=ab\x01
+        let mut field_soh = field.encode();
+        field_soh.push(SOH);
+
+        // encode the field into the message
+        message.extend_from_slice(field_soh.as_ref());
+    }
+
+    message
+}
+
+/// Prepends `8=BeginString` and `9=BodyLength` fields to the message buffer produced by `encode_regular_fields`.
```

Comment:
I would suggest avoiding the `message buffer produced by...` part. We don't really care about that, as we prepend to the message regardless of how the buffer was produced. I would go with: 


```suggestion
/// Prepends `8=BeginString` and `9=BodyLength` fields to the message buffer.
```